aten/src/ATen/native/LinearAlgebraUtils.h

```diff
@@ -385,27 +385,42 @@ static inline std::tuple<Tensor, Tensor, Tensor> _create_U_S_VT(const Tensor& in
   auto sizes = input.sizes().vec();
   int64_t m = input.size(-2), n = input.size(-1);
 
-  sizes[input.dim() - 1] = (compute_uv && some) ? std::min(m, n) : m;
-  auto strides = at::detail::defaultStrides(sizes);
+  sizes[input.dim() - 1] = some ? std::min(m, n) : m;
+  auto u_strides = at::detail::defaultStrides(sizes);
   // U should be a column-major or a batch of column-major matrices
   // ... x m x ucol will have strides: ...., ucol, 1
   // We require: ...., 1, m
-  strides[input.dim() - 1] = m;
-  strides[input.dim() - 2] = 1;
+  u_strides[input.dim() - 1] = m;
+  u_strides[input.dim() - 2] = 1;
 
-  Tensor U_empty = at::empty_strided(sizes, strides, input.options().device(usvt_device));
-  U_empty.zero_();
+  Tensor U_empty = compute_uv
+      ? at::empty_strided(sizes, u_strides, input.options().device(usvt_device))
+      : at::empty({0}, input.options().device(usvt_device));
 
   // VT should be a column-major or a batch of column-major matrices
-  sizes[input.dim() - 2] = n;
+  sizes[input.dim() - 2] = some ? std::min(m, n) : n;
   sizes[input.dim() - 1] = n;
-  // VT should be a column-major or a batch of column-major matrices
-  Tensor VT_empty = at::zeros(sizes, input.options().device(usvt_device));
-  VT_empty.transpose_(-2, -1);
+  auto vt_strides = at::detail::defaultStrides(sizes);
+  if (!svd_use_cusolver) {
+    vt_strides[input.dim() - 1] = sizes[input.dim() - 2];
+    vt_strides[input.dim() - 2] = 1;
+  }
+  Tensor VT_empty = compute_uv
+      ? at::empty_strided(sizes, vt_strides, input.options().device(usvt_device))
+      : at::empty({0}, input.options().device(usvt_device));
+
+  // U and VT might not get filled in this case
+  if (!some && compute_uv && input.numel() == 0) {
+    U_empty.zero_();
+    VT_empty.zero_();
+    // make U and VT an identity matrix, because they should be orthogonal
+    U_empty.diagonal(0, -2, -1).fill_(1);
+    VT_empty.diagonal(0, -2, -1).fill_(1);
```

Comment:
Would one `diag_embed` call work better here? `zero` and `fill` would invoke two cuda calls.

Reply:
This code is only for 0-sized inputs, that's a rare case.
`diag_embed` creates a tensor of zeros and copies provided tensor to the diagonal, so it's similar.
```cpp
  auto result = at::zeros(sizes, self.options());
  auto diag = result.diagonal(offset, dim1, dim2);
  diag.copy_(self);
```

Reply:
If `input.numel() == 0`, don't `U_empty` and `VT_empty` also have zero elements?

Reply:
No, for example when input has size 0x3. U would have size 0x0, and V - 3x3.

Reply:
Ah, I see. Then it should be rather than `numel() == 0`, `size(-2) == 0 || size(-1) == 0`, right?

Reply:
The logic here is correct afau, if input.numel() == 0 `u` and `v` may or may not be empty (but they will be correctly sized regardless), and if they are not, by convention we are making them (a batch of) identity matrices so that they are orthogonal. If they are empty, no harm done, `zero_`, `diagonal` and `fill_` work on empty tensors.